### DISCUSSION TITLE
Should extracting a variable with Concatenate trigger an incompatible override error?

### DISCUSSION THREAD

**Discussion by cdce8p** [microsoft/pyright#2844] [General]
With v1.1.206, the example below now triggers an incompatibleOverride error. I guess something that was hidden before #2838. I'm not sure what the best course of action here should be. Just by definition of PEP 612 the error is correct. `self` in `wrapper` is defined as `positional-only`. In practice however it would be unnecessary cumbersome to have to add `/` after any `self`.  I might have mentioned it already elsewhere, but this is also something that's causing us headaches at the mypy side. https://github.com/python/mypy/pull/11847

Any ideas how to best resolve this?

```py
from __future__ import annotations
from typing import Any, Callable, Concatenate, ParamSpec, TypeVar

T = TypeVar("T", bound="Parent")
R = TypeVar("R")
P = ParamSpec("P")


def decorator(
    func: Callable[Concatenate[T, P], R]
) -> Callable[Concatenate[T, P], R]:
    def wrapper(self: T, *args: P.args, **kwargs: P.kwargs) -> R:
        try:
            return func(self, *args, **kwargs)
        except Exception:
            self.error_handling()
            raise
    return wrapper

class Parent:
    def error_handling(self) -> None:
        ...

    def func(self, y: int, **kwargs: Any) -> None:
        ...

class Child(Parent):
    @decorator
    def func(self, y: int, **kwargs: Any) -> None:
        ...

reveal_type(Parent.func)
reveal_type(Child.func)
```

The fallback would probably be to use `cast`. The downside being that the decorator could now be used everywhere since `self` isn't checked anymore.
```py
def decorator(
    func: Callable[P, R]
) -> Callable[P, R]:
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
        self = cast(Parent, args[0])
        try:
            return func(*args, **kwargs)
        except Exception:
            self.error_handling()
            raise
    return wrapper
```

**Comment by erictraut** [comment]
The original method has a parameter `self` that can be specified as a positional or keyword argument. The transformed version accepts only a positional argument. You can eliminate this by adding a positional-only separator after the `self` parameter in the base class method.

```python
    def func(self, /, y: int, **kwargs: Any) -> None:
        ...
```

The error message is far from clear, so I'll file a bug to improve that.

I'll also think about whether this could be special-cased because others are likely to hit this issue.

**Reply by cdce8p** [reply]
Thanks for the response! Adding `/` to `Parent.func` is what I meant with "unnecessary cumbersome" 😄
So yeah personally, I think it would make sense to special case it.

I'm thinking about if I missed something in how to type this pattern, but keep coming back to the double `Concatenate`. Except for the `positional-only` headaches, it does exactly what I would want it to.
